core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -156,29 +166,30 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
   final def raceWith[E1, E2, B, C](
     that: IO[E1, B]
   )(
-    leftWins: (Fiber[E, A], Fiber[E1, B]) => IO[E2, C],
-    rightWins: (Fiber[E1, B], Fiber[E, A]) => IO[E2, C]
+    leftDone: (ExitResult[E, A], Fiber[E1, B]) => IO[E2, C],
+    rightDone: (ExitResult[E1, B], Fiber[E, A]) => IO[E2, C]
   ): IO[E2, C] = {
     def arbiter[E0, E1, A, B](
-      f: (Fiber[E0, A], Fiber[E1, B]) => IO[E2, C],
-      winner: Fiber[E0, A],
+      f: (ExitResult[E0, A], Fiber[E1, B]) => IO[E2, C],
       loser: Fiber[E1, B],
-      race: Ref[IO.Race],
+      race: Ref[Int],
       done: Promise[E2, C]
     )(res: ExitResult[E0, A]): IO[Nothing, _] =
       race
-        .modify(r => r.won(res.succeeded) -> r.next(res.succeeded))
-        .flatMap(IO.when(_)(f(winner, loser).to(done).void))
-
-    (for {
-      done  <- Promise.make[E2, C]
-      race  <- Ref[IO.Race](IO.Race.Started)
-      left  <- self.fork
-      right <- that.fork
-      _     <- left.observe.flatMap(arbiter(leftWins, left, right, race, done)).fork
-      _     <- right.observe.flatMap(arbiter(rightWins, right, left, race, done)).fork
-      c     <- done.get
-    } yield c).supervised
+        .modify((c: Int) => (if (c > 0) IO.unit else f(res, loser).to(done).void) -> (c + 1))
+        .flatMap(identity(_))
+
+    for {
+      done <- Promise.make[E2, C]
+      race <- Ref[Int](0)
+      _ <- (for {
+            left  <- self.fork
+            right <- that.fork
+            _     <- left.observe.flatMap(arbiter(leftDone, right, race, done)).fork
+            _     <- right.observe.flatMap(arbiter(rightDone, left, race, done)).fork
+          } yield ()).uninterruptibly
```

Comment:
I can't decide. I _think_ what we need to do is `onInterrupt` of the parent, kill the children. This way the children can outlive the parent under normal circumstances. But if the parent is interrupted, the forked fibers are cleaned up.

Some things would be faster (performance wise) if you could check to see if you're interrupted. Then you could just stuff something in `ensuring` and it would check the interruption status. Maybe `FiberDescriptor` can help here.

Reply:
Yeah, it's already a boolean in `FiberStatus`, easy to surface. Given its special meaning, I think `Cause.Interrupted` should be promoted to something else; right now it's lost in a sea of other errors, but it's not an error, it's a flag really (`isInterrupted`).

Reply:
It's true, but in the model that every fiber must `Succeed` or `Fail`, one of the reasons it can fail is interruption. Also when we start composing `Cause`, we need to preserve how fibers failed: maybe one fiber failed due to interruption, but another failed due to an `E`. So this means interruption must be preserved inside `Cause`. 

Finally interruption occurs at a certain type. If, for example, a fiber fails with `E` but then is interrupted while it's unwinding the stack and running finalizers, then the `Cause` should reflect that: the primary failure with `E` _followed by_ an interruption. I think all that works properly, or is at least very close to working properly.

All right, I'll modify descriptor to contain an interrupted flag because it will make some things _way_ faster.